### PR TITLE
More SyntaxContext sharing among CompletionProviders

### PR DESCRIPTION
**Pull request by genlu**
Followup on #61057

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs [hunk 1]
```diff
@@ -53,14 +53,15 @@ public override async Task ProvideCompletionsAsync(CompletionContext completionC
...
-                var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
 
                 if (!completionContext.CompletionOptions.ShowNameSuggestions)
                 {
                     return;
                 }
 
-                var context = CSharpSyntaxContext.CreateContext(document, semanticModel, position, cancellationToken);
+                var context = await completionContext.GetSyntaxContextWithExistingSpeculativeModelAsync(document, cancellationToken).ConfigureAwait(false) as CSharpSyntaxContext;
+                Contract.ThrowIfNull(context);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:63]: do we need this contract call?  how can this fail or return null?

File: src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs [hunk 2]
```diff
@@ -36,15 +36,15 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
...
         }
 
-        private async Task<ImmutableArray<CompletionItem>> RecommendCompletionItemsAsync(Document document, int position, CancellationToken cancellationToken)
+        private async Task<ImmutableArray<CompletionItem>> RecommendCompletionItemsAsync(Document document, CompletionContext context, CancellationToken cancellationToken)
         {
-            var syntaxContextService = document.GetRequiredLanguageService<ISyntaxContextService>();
-            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
-            var syntaxContext = (TContext)syntaxContextService.CreateContext(document, semanticModel, position, cancellationToken);
+            var position = context.Position;
+            var syntaxContext = await context.GetSyntaxContextWithExistingSpeculativeModelAsync(document, cancellationToken).ConfigureAwait(false) as TContext;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:46]:
```suggestion
            var syntaxContext = (TContext)await context.GetSyntaxContextWithExistingSpeculativeModelAsync(document, cancellationToken).ConfigureAwait(false);
```

File: src/Workspaces/Core/Portable/Recommendations/IRecommendationService.cs [hunk 3]
```diff
@@ -16,6 +17,11 @@ internal interface IRecommendationService : ILanguageService
             int position,
             RecommendationServiceOptions options,
             CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Recommendations/IRecommendationService.cs:19]: do we need thsi entrypoint anymore?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs (hunk 1 above)
Comment (genlu) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:63]: Just to make nullable analysis happy

File: src/Workspaces/Core/Portable/Recommendations/IRecommendationService.cs (hunk 3 above)
Comment (genlu) [reply src/Workspaces/Core/Portable/Recommendations/IRecommendationService.cs:19]: Yes, there are callers other than completion

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs:63]: i'm not sure why any of this stuff is annotated as returning null though...

**Comment by genlu** [comment]
@CyrusNajmabadi Good to merge?

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**